packages/fullscreen/src/index.ts

```diff
@@ -0,0 +1,154 @@
+import { useState } from "react";
+
+type EventCallback = (this: Document, ev: any) => any;
+interface NormalizedFullscreenApi {
+  requestFullscreen: string;
+  exitFullscreen: string;
+  fullscreenElement: string;
+  fullscreenEnabled: string;
+  fullscreenchange: string;
+  fullscreenerror: string;
+}
+
+const getBrowserFunctions = (): NormalizedFullscreenApi => {
+  let val;
+
+  const fnMap = [
+    [
+      "requestFullscreen",
+      "exitFullscreen",
+      "fullscreenElement",
+      "fullscreenEnabled",
+      "fullscreenchange",
+      "fullscreenerror"
+    ],
+    // New WebKit
+    [
+      "webkitRequestFullscreen",
+      "webkitExitFullscreen",
+      "webkitFullscreenElement",
+      "webkitFullscreenEnabled",
+      "webkitfullscreenchange",
+      "webkitfullscreenerror"
+    ],
+    // Old WebKit
+    [
+      "webkitRequestFullScreen",
+      "webkitCancelFullScreen",
+      "webkitCurrentFullScreenElement",
+      "webkitCancelFullScreen",
+      "webkitfullscreenchange",
+      "webkitfullscreenerror"
+    ],
+    [
+      "mozRequestFullScreen",
+      "mozCancelFullScreen",
+      "mozFullScreenElement",
+      "mozFullScreenEnabled",
+      "mozfullscreenchange",
+      "mozfullscreenerror"
+    ],
+    [
+      "msRequestFullscreen",
+      "msExitFullscreen",
+      "msFullscreenElement",
+      "msFullscreenEnabled",
+      "MSFullscreenChange",
+      "MSFullscreenError"
+    ]
+  ];
+
+  let i = 0;
+  const l = fnMap.length;
+  const ret = {} as NormalizedFullscreenApi;
+
+  for (; i < l; i++) {
+    val = fnMap[i];
+    if (val && val[1] in document) {
+      for (i = 0; i < val.length; i++) {
+        ret[fnMap[0][i]] = val[i];
+      }
+      return ret;
+    }
+  }
+};
+
+const useFullscreen = () => {
+  const [isFullscreen, setIsFullscreen] = useState(false);
+  const [element, setElement] = useState(null);
+  const fn = getBrowserFunctions();
+
+  const eventNameMap = {
+    change: fn.fullscreenchange,
+    error: fn.fullscreenerror
+  };
+
+  const request = (element?: HTMLElement) =>
+    new Promise((resolve, reject) => {
+      const onFullScreenEntered = () => {
+        off("change", onFullScreenEntered);
+        resolve();
+      };
+
+      on("change", onFullScreenEntered);
+
+      element = element || document.documentElement;
+      setIsFullscreen(true);
+      setElement(element);
+
+      Promise.resolve(element[fn.requestFullscreen]()).catch(reject);
+    });
+
+  const on = (event: string, callback: EventCallback) => {
+    const eventName = eventNameMap[event];
+    if (eventName) {
+      document.addEventListener(eventName, callback, false);
+    }
+  };
+  const off = (event: string, callback: EventCallback) => {
+    const eventName = eventNameMap[event];
+    if (eventName) {
+      document.removeEventListener(eventName, callback, false);
+    }
+  };
+  const exit = () =>
+    new Promise(function(resolve, reject) {
```

Comment:
Can you make sure you're consistent with functions? It's better to use arrow functions in callbacks, as the can be anonymous.

Reply:
Good call, thank you for pointing this out.